### ISSUE TITLE
OSError: Can't read some data (inflate() failed) & (wrong B-tree signature)

### ISSUE THREAD

**Issue by fprada** [vaexio/vaex#625]
Hi,

  we have succeed to read using vaex about 40TB of HDF5 files. It is very remarkable how fast it does work vaex, congratulations!

  Each directory correspond to a simulation epoch, there are 50 of them, and inside each directory we do have 100 files (from 0 to 99). Vaex works perfectly well, i.e. 

  ds = vaex.open('./halodir_025/halolist_*.h5')

 BUT, it cannot read the data for halodir_030 and halodir_032 for some reason. I have tested that each of the 100 files inside each directory can be read with vaex, but not at once. I noticed that in halodir_030 vaex crashes when I read halolist_8*.h5 but it works if I read every other group of tenths. e.g. halolist_3*.h5, and even if I read each of the 8* individually (81, 82, ... 89) it works. For halodir_032 the problem is similar, it is happening for 1* and 4* ... Yet, the error message I get seems different: for halodir_030 I get "OSError: Can't read some data (inflate() failed)" and for halodir_032 I get "OSError: Can't read some data (wrong B-tree signature)". LIKELY because vaex cannot read all 100 files at once, as shown above for halodir_025.

Below I include the full errors for both cases:

**halodir_030**
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-105-44c00e23cd35> in <module>
----> 1 ds = vaex.open('/home/users/dae/ishiyama/Uchuu/RockstarExtendedAll/RockstarExtended/halodir_030/halolist_*.h5')

~/.conda/envs/vaexenv/lib/python3.7/site-packages/vaex/__init__.py in open(path, convert, shuffle, copy_index, *args, **kwargs)
    211                     for filename in filenames:
    212                         DataFrames.append(open(filename, convert=bool(convert), shuffle=shuffle, **kwargs))
--> 213                     ds = vaex.dataframe.DataFrameConcatenated(DataFrames)
    214                 if convert:
    215                     ds.export_hdf5(filename_hdf5, shuffle=shuffle)

~/.conda/envs/vaexenv/lib/python3.7/site-packages/vaex/dataframe.py in __init__(self, dfs, name)
   5277         self.columns = {}
   5278         for column_name in self.get_column_names(virtual=False):
-> 5279             self.columns[column_name] = ColumnConcatenatedLazy(dfs, column_name)
   5280             self._save_assign_expression(column_name)
   5281 

~/.conda/envs/vaexenv/lib/python3.7/site-packages/vaex/column.py in __init__(self, dfs, column_name)
    102         self.dfs = dfs
    103         self.column_name = column_name
--> 104         dtypes = [df.dtype(column_name) for df in dfs]
    105         self.is_masked = any([df.is_masked(column_name) for df in dfs])
    106         if self.is_masked:

~/.conda/envs/vaexenv/lib/python3.7/site-packages/vaex/column.py in <listcomp>(.0)
    102         self.dfs = dfs
    103         self.column_name = column_name
--> 104         dtypes = [df.dtype(column_name) for df in dfs]
    105         self.is_masked = any([df.is_masked(column_name) for df in dfs])
    106         if self.is_masked:

~/.conda/envs/vaexenv/lib/python3.7/site-packages/vaex/dataframe.py in dtype(self, expression, internal)
   1920         elif expression in self.columns.keys():
   1921             column = self.columns[expression]
-> 1922             data = column[0:1]
   1923             dtype = data.dtype
   1924         else:

~/.conda/envs/vaexenv/lib/python3.7/site-packages/vaex/column.py in __getitem__(self, slice)
     62 
     63     def __getitem__(self, slice):
---> 64         return self.ar[slice]
     65 
     66     def __setitem__(self, slice, value):

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

~/.conda/envs/vaexenv/lib/python3.7/site-packages/h5py/_hl/dataset.py in __getitem__(self, args)
    571         mspace = h5s.create_simple(mshape)
    572         fspace = selection.id
--> 573         self.id.read(mspace, fspace, arr, mtype, dxpl=self._dxpl)
    574 
    575         # Patch up the output for NumPy

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/h5d.pyx in h5py.h5d.DatasetID.read()

h5py/_proxy.pyx in h5py._proxy.dset_rw()

h5py/_proxy.pyx in h5py._proxy.H5PY_H5Dread()

OSError: Can't read data (inflate() failed)

**halodir_032**
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-106-16934bec277e> in <module>
----> 1 ds = vaex.open('/home/users/dae/ishiyama/Uchuu/RockstarExtendedAll/RockstarExtended/halodir_032/halolist_*.h5')

~/.conda/envs/vaexenv/lib/python3.7/site-packages/vaex/__init__.py in open(path, convert, shuffle, copy_index, *args, **kwargs)
    211                     for filename in filenames:
    212                         DataFrames.append(open(filename, convert=bool(convert), shuffle=shuffle, **kwargs))
--> 213                     ds = vaex.dataframe.DataFrameConcatenated(DataFrames)
    214                 if convert:
    215                     ds.export_hdf5(filename_hdf5, shuffle=shuffle)

~/.conda/envs/vaexenv/lib/python3.7/site-packages/vaex/dataframe.py in __init__(self, dfs, name)
   5277         self.columns = {}
   5278         for column_name in self.get_column_names(virtual=False):
-> 5279             self.columns[column_name] = ColumnConcatenatedLazy(dfs, column_name)
   5280             self._save_assign_expression(column_name)
   5281 

~/.conda/envs/vaexenv/lib/python3.7/site-packages/vaex/column.py in __init__(self, dfs, column_name)
    102         self.dfs = dfs
    103         self.column_name = column_name
--> 104         dtypes = [df.dtype(column_name) for df in dfs]
    105         self.is_masked = any([df.is_masked(column_name) for df in dfs])
    106         if self.is_masked:

~/.conda/envs/vaexenv/lib/python3.7/site-packages/vaex/column.py in <listcomp>(.0)
    102         self.dfs = dfs
    103         self.column_name = column_name
--> 104         dtypes = [df.dtype(column_name) for df in dfs]
    105         self.is_masked = any([df.is_masked(column_name) for df in dfs])
    106         if self.is_masked:

~/.conda/envs/vaexenv/lib/python3.7/site-packages/vaex/dataframe.py in dtype(self, expression, internal)
   1920         elif expression in self.columns.keys():
   1921             column = self.columns[expression]
-> 1922             data = column[0:1]
   1923             dtype = data.dtype
   1924         else:

~/.conda/envs/vaexenv/lib/python3.7/site-packages/vaex/column.py in __getitem__(self, slice)
     62 
     63     def __getitem__(self, slice):
---> 64         return self.ar[slice]
     65 
     66     def __setitem__(self, slice, value):

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

~/.conda/envs/vaexenv/lib/python3.7/site-packages/h5py/_hl/dataset.py in __getitem__(self, args)
    571         mspace = h5s.create_simple(mshape)
    572         fspace = selection.id
--> 573         self.id.read(mspace, fspace, arr, mtype, dxpl=self._dxpl)
    574 
    575         # Patch up the output for NumPy

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/_objects.pyx in h5py._objects.with_phil.wrapper()

h5py/h5d.pyx in h5py.h5d.DatasetID.read()

h5py/_proxy.pyx in h5py._proxy.dset_rw()

h5py/_proxy.pyx in h5py._proxy.H5PY_H5Dread()

OSError: Can't read data (wrong B-tree signature)

**Comment by maartenbreddels** [comment]
Hi,

great to hear you like it :)
I think your hdf5 file is corrupted, can you try some of the h5 command-line utilities on it? h5dump/h5ls/h5copy

cheers,

Maarten

**Comment by fprada** [comment]
Thanks Maarten!

   We've checked and those hdf5 files are perfectly fine when we read then with python by using h5py.file (import h5py).  We computed the Halo Mass Function for those 2 directories (redshifts) and results are perfectly fine.

   Btw. h5ls works perfectly fine on them.

Best wishes,
 Paco.

**Comment by fprada** [comment]
Hi again,

  any idea of what could happen?

Thanks!

**Comment by LeeKamentsky** [comment]
I found this while googling. I've experienced similar. In my case, I was using Python's multiprocessing library to read chunks out of an HDF5 file using h5py, opening the file in read-only mode and still getting this error.

I think the problem is that when h5py opens the file handles in the parent process, somehow they don't survive the trip to the child process in multiprocessing. My solution was to prevent the parent process from opening the file, opening it in each child process instead, e.g. in this code snippet: https://github.com/chunglabmit/spimstitch/blob/d5e25e9148d7f41182c2b97253e8ef06a973e3f6/spimstitch/imaris.py#L70
